src/auth/auth_gitlab.py

```diff
@@ -0,0 +1,339 @@
+        self.user_states[user]['updated'] = now
+        self.user_states[user]['visit'] = now
+        self.clean_expired_sessions()
+        self.dump_sessions_to_file()
+        return
+
+    @gen.coroutine
+    def update_user_state(self, user, access_token):
+        user_state = yield self.fetch_user(access_token)
+        if user_state is None:
+            LOGGER.error("Failed to fetch user %s" % user)
+            self.user_states[user]['state'] = "error"
+        else:
+            LOGGER.info("User %s refreshed: " % user + str(user_state))
+            self.user_states[user] = {**self.user_states[user], **user_state}
+        now = time.time()
+        self.user_states[user]['updating'] = False
+        self.user_states[user]['updated'] = now
+        self.user_states[user]['visit'] = now
+        self.clean_expired_sessions()
+        self.dump_sessions_to_file()
+        return
+
+    @gen.coroutine
+    def read_groups(self, access_token):
+        args = {
+            'access_token': access_token,
+            'all_available': 'false',
+            'per_page': 100,
+        }
+        if not self.gitlab_group_search is None:
+            args['search'] = self.gitlab_group_search
+
+        group_list_future = self.oauth2_request(
+            self._OAUTH_GITLAB_GROUPS % self._GITLAB_PREFIX,
+            **args
+        )
+
+        group_list = yield group_list_future
+
+        if group_list is None:
+            return None
+
+        groups = []
+        for group in group_list:
+            if group.get('full_path'):
+                groups.append(group['full_path'])
+
+        return groups
+
+    @gen.coroutine
+    def validate_user(self, code, request_handler):
+        user_response_future = self.get_authenticated_user(
+            get_path_for_redirect(request_handler),
+            self.client_id,
+            self.secret,
+            code
+        )
+        user_response = yield user_response_future
+
+        if user_response.get('email') is None:
+            error_message = 'No email field in user response. The response: ' + str(user_response)
+            LOGGER.error(error_message)
+            raise AuthFailureError(error_message)
+
+        user_groups = []
+        if self.gitlab_group_support:
+            user_groups = yield self.read_groups(user_response.get('access_token'))
+            if user_groups is None:
+                error_message = 'Cant read user groups'
+                LOGGER.error(error_message)
+                raise AuthFailureError(error_message)
+
+        LOGGER.info("User %s group list: " % user_response['email'] + str(user_groups))
+        user_response['groups'] = user_groups
+        user_response['updated'] = time.time()
+        user_response['visit'] = time.time()
+        user_response['updating'] = False
+        oauth_access_token = user_response.pop('access_token')
+        oauth_refresh_token = user_response.pop('refresh_token') # not used atm
```

Comment:
If it's not used, could you remove it, please? Or you are going to add it later in PR?

Reply:
Yep, I can remove. I just dont known if refresh token really works for gitlab. Anyway any refreshing of expired token not implemented, so can remove atm